### ISSUE TITLE
parseJSON parses JSON String with invalid Date to valid but different Date

### ISSUE THREAD

**Issue by Schmaga** [date-fns/date-fns#2517]
Hello everyone, I was in the process of migrating current logic from moment.js to date-fns in a project when I bumped into the following behavior.

Browser: Latest Chrome (91)

I pass in an invalid date but as a syntactically correct JSON date string and it returns a date object with a different date:
```ts
// returns the first of March as valid date: (Sun Mar 01 2020 01:00:00 GMT+0100)
const parsedDate = parseJSON('2020-02-30T00:00:00+00:00');
isValid(parsedDate) // returns true

// as reference, this was the previous moment.js based implementation which correctly found it an invalid date:
const utcMoment = moment.utc('2020-02-30T00:00:00+00:00');
utcMoment.isValid() // returns false
```

The code is used to validate dates that users can enter.  Luckily I had tests for edge cases of invalid dates. Interestingly though, JavaScript `Date` shows the same behavior:
```ts
// also returns the first of March as valid date: (Sun Mar 01 2020 01:00:00 GMT+0100)
const parsedDate = new Date('2020-02-30T00:00:00+00:00');
```

In 2020 the 29th of February was a valid date, so I just assume that the parsing adds the difference in days on top and ends up on the first of March. It might also have to do with my timezone (GMT + 1).

I am not sure if this is by design, because when referring to the documentation, I would have expected an "Invalid Date". Is this a bug or am I using the function wrongly?

**Comment by fturmel** [comment]
Try using the `parseISO` function instead. I did a quick test and it does return an invalid date for 2020-02-30 as expected.

`parseJSON` is very lean and really just splits the string into different parts to create a native Date object. It would be roughly equivalent to `new Date(Date.UTC(2020, 1, 30))` here.

https://github.com/date-fns/date-fns/blob/master/src/parseJSON/index.ts#L38-L62

**Comment by Schmaga** [comment]
Using `parseISO` seems to help with that specific case. Thanks!

But I guess my original find is still a bug?

Also, the `parse` function I use for locale-specific parsing seems to have the same problem. See this example:
```ts
// this returns a date of 01.03.2019 in german locale
const parsedDate = parse('29.12.2019', 'P', new Date(), {
      locale: this.currentLocale, // german locale
    });
```

**Comment by fturmel** [comment]
As for the locale `parse` issue, I can't reproduce it on my end. I am getting an invalid date for your example when I recreate it.

Have a look at this demo, does it work for you?
https://codesandbox.io/s/vigilant-blackwell-qremn?file=/src/index.js

**Comment by Schmaga** [comment]
> As for the locale `parse` issue, I can't reproduce it on my end. I am getting an invalid date for your example when I recreate it.
> 
> Have a look at this demo, does it work for you?
> https://codesandbox.io/s/vigilant-blackwell-qremn?file=/src/index.js

You are right. It now works for me as well. I think I might have looked at the wrong failing code during testing. So it boils down to the initial parseJSON issue. I will leave this open to give the maintainers a chance to look at and clarify.

**Comment by robations** [comment]
This could be handled in documentation perhaps?

> Any other input type or invalid date strings will return an Invalid Date.

https://date-fns.org/v2.29.3/docs/parseJSON

This probably means invalid from a syntax POV. Could perhaps state that trusted inputs are expected and this is a performance trade-off.
